LedgerHQ/ledger-live
# [LIVE-11369][LLM] Delete feature flag "llmNewFirmwareUpdateUx", replace by business logic

### 📝 Description

Right now, the `llmNewFirmwareUpdateUx` feature flag enables the new firmware update UX for ALL devices at once (including Nanos over USB, not for Nano X BLE). It is inconvenient as this flow needs adaptations in order to work properly on Nano devices. **The feature flag therefore serves no purpose**.

- Functional changes:
  - Delete the flag `llmNewFirmwareUpdateUx`
  - Refactor the decisional tree so that the new update flow is only available for Stax & Europa (BLE+USB) (for now) and that for other (wired) devices it uses the old fully functional flow, as it is today in production.
  - Fix end of firmware update: use `setLastSeenDeviceInfo` (instead of broken `setLastSeenDevice`) so that LLM knows the version of the last seen device. The `setLastSeenDevice` action creator had no associated reducer so it was not doing anything at all. Because of that, after a first successful update, the firmware update banner was not disappearing (or getting updated info in case of another update).
- Refactoring:
  - The entry point for the firmware update is the component `FirmwareUpdateBanner` and the logic it contains
    - Moved this component to `newArch`.
    - Split the logic in smaller logical blocks with unit tests

**new tests:**
![Screenshot 2024-03-01 at 13 28 21](https://github.com/LedgerHQ/ledger-live/assets/91890529/592fadb9-59b7-4fc4-8695-28568064055f)
![Screenshot 2024-03-04 at 16 51 18](https://github.com/LedgerHQ/ledger-live/assets/91890529/d15427ec-2e74-477e-817d-da8e14ce492f)



**QA: what should be tested ?**
- For the last connected device, if a firmware update is NOT available, the firmware update banner should not be displayed on the portfolio page or in My Ledger.
- For the last connected device, if a firmware update is available, the firmware update banner should be displayed on the portfolio page and in My Ledger.
- Tapping on the firmware update banner:
  - For Ledger Stax, it should trigger a navigation to the new firmware update flow. This is the flow were the UI shows the several steps of the update.
  - For Ledger Nano models:
    - iOS, any device model: it should display a banner prompting the user to update their device using LLD.
    - Android, Nano X that was last connected via bluetooth: it should prompt the user to update their device using LLD.
    - Android, Nano (S/SP/X) that was last connected via USB: if the version does not satisfy this criteria, it should display a banner prompting the user to update their device using LLD:
      - Nano S: version ≥ 1.6.1
      - Nano X: version ≥ 1.3.0
      - Nano SP: version ≥ 1.0.0
    - Android, Nano (S/SP/X) last connected via USB: if the version satisfies the criteria above, it should navigate to the (old) firmware update flow.


### ❓ Context

- **JIRA or GitHub link**: [LIVE-11369]

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

  - LLM: firmware update banner and redirection to firmware update flows depending on the device, its firmware version and the mobile OS.

---

### 🧐 Checklist for the PR Reviewers



- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)


[LIVE-11369]: https://ledgerhq.atlassian.net/browse/LIVE-11369?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
